File: web-site/src/lib/achievements.js
// Conquistas (medalhas). Iguais no app (lib/services/achievements.dart).
// São calculadas a partir dos dados da conta; `stats` guarda os contadores:
//   stats: { correct, bestStreak, rankedGames, dailyDone, dailyPerfect,
//            lastDaily: 'AAAA-MM-DD', days: ['AAAA-MM-DD'], weeks: ['AAAA-MM-DD'] }

export const EMPTY_STATS = {
  correct: 0,
  bestStreak: 0,
  rankedGames: 0,
  dailyDone: 0,
  dailyPerfect: 0,
  lastDaily: null,
  days: [],
  weeks: [],
}

const fullTeam = (t) => (t.pokemon ?? []).filter((p) => p != null).length === 6

export const ACHIEVEMENTS = [
  { id: 'first', icon: '🎯', title: 'Primeiro acerto', text: 'Acerte um Pokémon no jogo.', done: (s) => s.stats.correct >= 1 },
  { id: 'trainer', icon: '🎒', title: 'Treinador', text: 'Acerte 100 Pokémon no jogo.', done: (s) => s.stats.correct >= 100 },
  { id: 'master', icon: '🎓', title: 'Mestre Pokémon', text: 'Acerte 1000 Pokémon no jogo.', done: (s) => s.stats.correct >= 1000 },
  { id: 'streak', icon: '🔥', title: 'Em chamas', text: 'Faça uma sequência de 10 acertos.', done: (s) => s.stats.bestStreak >= 10 },
  { id: 'unstoppable', icon: '☄️', title: 'Imparável', text: 'Faça uma sequência de 50 acertos.', done: (s) => s.stats.bestStreak >= 50 },
  { id: 'ranked', icon: '🏆', title: 'Competidor', text: 'Faça 100 pontos no Ranked.', done: (s) => s.rankedRecord >= 100 },
  { id: 'lightning', icon: '⚡', title: 'Relâmpago', text: 'Chegue a 400 pontos no Ranked (2 s por Pokémon).', done: (s) => s.rankedRecord >= 400 },
  { id: 'daily', icon: '📅', title: 'Desafiante', text: 'Complete um desafio do dia.', done: (s) => s.stats.dailyDone >= 1 },
  { id: 'loyal', icon: '🗓️', title: 'Fiel', text: 'Complete 7 desafios do dia.', done: (s) => s.stats.dailyDone >= 7 },
  { id: 'perfect', icon: '💯', title: 'Perfeito', text: 'Acerte os 10 Pokémon de um desafio do dia.', done: (s) => s.stats.dailyPerfect >= 1 },
  { id: 'collector', icon: '⭐', title: 'Colecionador', text: 'Tenha 10 Pokémon favoritos.', done: (s) => s.favorites.length >= 10 },
  { id: 'team', icon: '🛡️', title: 'Time completo', text: 'Monte um time com 6 Pokémon.', done: (s) => s.teams.some(fullTeam) },
  { id: 'strategist', icon: '♟️', title: 'Estrategista', text: 'Tenha 5 times completos.', done: (s) => s.teams.filter(fullTeam).length >= 5 },
]

/** Lista de conquistas com `unlocked` para o estado atual. */
export function achievementsOf({ stats, rankedRecord = 0, favorites = [], teams = [] }) {
  const state = { stats: { ...EMPTY_STATS, ...(stats ?? {}) }, rankedRecord, favorites, teams }
  return ACHIEVEMENTS.map((a) => ({ ...a, unlocked: a.done(state) }))
}

File: lib/services/achievements.dart
// lib/services/achievements.dart
//
// Conquistas (medalhas) — as mesmas do site (web-site/src/lib/achievements.js).
// São calculadas a partir dos dados da conta; `stats` guarda os contadores:
//   stats: {correct, bestStreak, rankedGames, dailyDone, dailyPerfect,
//           lastDaily: 'AAAA-MM-DD', days: [...], weeks: [...]}

class Achievement {
  final String id;
  final String icon;
  final String title;
  final String text;
  final bool unlocked;
  const Achievement(this.id, this.icon, this.title, this.text, this.unlocked);
}

class Achievements {
  Achievements._();

  static Map<String, dynamic> emptyStats() => {
        'correct': 0,
        'bestStreak': 0,
        'rankedGames': 0,
        'dailyDone': 0,
        'dailyPerfect': 0,
        'lastDaily': null,
        'days': <dynamic>[],
        'weeks': <dynamic>[],
      };

  static int _n(Map<String, dynamic> stats, String key) => (stats[key] as num?)?.toInt() ?? 0;

  static bool _full(Map<String, dynamic> team) =>
      ((team['pokemon'] as List?) ?? []).where((p) => p != null).length == 6;

  static List<Achievement> of({
    required Map<String, dynamic> stats,
    required int rankedRecord,
    required List<int> favorites,
    required List<Map<String, dynamic>> teams,
  }) {
    final s = {...emptyStats(), ...stats};
    final fullTeams = teams.where(_full).length;
    return [
      Achievement('first', '🎯', 'Primeiro acerto', 'Acerte um Pokémon no jogo.', _n(s, 'correct') >= 1),
      Achievement('trainer', '🎒', 'Treinador', 'Acerte 100 Pokémon no jogo.', _n(s, 'correct') >= 100),
      Achievement('master', '🎓', 'Mestre Pokémon', 'Acerte 1000 Pokémon no jogo.', _n(s, 'correct') >= 1000),
      Achievement('streak', '🔥', 'Em chamas', 'Faça uma sequência de 10 acertos.', _n(s, 'bestStreak') >= 10),
      Achievement('unstoppable', '☄️', 'Imparável', 'Faça uma sequência de 50 acertos.', _n(s, 'bestStreak') >= 50),
      Achievement('ranked', '🏆', 'Competidor', 'Faça 100 pontos no Ranked.', rankedRecord >= 100),
      Achievement('lightning', '⚡', 'Relâmpago', 'Chegue a 400 pontos no Ranked (2 s por Pokémon).', rankedRecord >= 400),
      Achievement('daily', '📅', 'Desafiante', 'Complete um desafio do dia.', _n(s, 'dailyDone') >= 1),
      Achievement('loyal', '🗓️', 'Fiel', 'Complete 7 desafios do dia.', _n(s, 'dailyDone') >= 7),
      Achievement('perfect', '💯', 'Perfeito', 'Acerte os 10 Pokémon de um desafio do dia.', _n(s, 'dailyPerfect') >= 1),
      Achievement('collector', '⭐', 'Colecionador', 'Tenha 10 Pokémon favoritos.', favorites.length >= 10),
      Achievement('team', '🛡️', 'Time completo', 'Monte um time com 6 Pokémon.', fullTeams >= 1),
      Achievement('strategist', '♟️', 'Estrategista', 'Tenha 5 times completos.', fullTeams >= 5),
    ];
  }
}
